### PR TITLE
chore(facade): Introduce AwaitCurrentDispatches

### PR DESCRIPTION
**Pull request by dranikpg**
We have some places in our code base that first push a global change and then need to wait until there are no commands anymore running that have values of the old state cached / are amidst a breaking execution:

* When doing takeover, we first set the global state to taking-over and then wait for ongoing commands to finish. No new commands will be able to run because of the global state rejecting all commands
* When pushing cluster configs, before we delete moved slots, we want to make sure no command is accessing them, see #1480 
* Possibly when doing client pause (currently a PR)

The current implementation of `AwaitDispatches` waits until a point of time when _no dispatches at all_ are happening - which is too strong of a guarantee, we only want to wait for commands that were amidst execution during the update to finish. 

The first case is lucky as is disallows execution at all, so there are no suspension points in the commands, meaning the current AwaitDispatches loop will only execute in-between dispatches. However with cluster config changes, no commands are rejected, so even under moderate load there just may be no such moment that no command is currently running -> means AwaitDispatches will always fail

The new implementation of `AwaitCurrentDispatches` just enqueues checkpoints for all connections which are currently dispatching. Once they're done, they reach the checkpoint and decrement the counter

### CODE REVIEW

File: src/facade/dragonfly_connection.cc [hunk 1]
```diff
@@ -190,35 +191,32 @@ size_t Connection::PipelineMessage::StorageCapacity() const {
...
+    }
+    size_t operator()(const MonitorMessage& msg) {
+      return msg.capacity();
+    }
+    size_t operator()(const AclUpdateMessage& msg) {
+      return 0;
+    }
+    size_t operator()(const MigrationRequestMessage& msg) {
+      return 0;
+    }
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:211]: Intentional: this is the heap size _only_ (should add a comment), because the size of the variant itself is included below

File: src/facade/dragonfly_connection.cc [hunk 2]
```diff
@@ -1143,32 +1153,21 @@ void Connection::SendAsync(MessageHandle msg) {
...
+      ++it;
+    dispatch_q_.insert(it, std::move(msg));
   } else {
     dispatch_q_.push_back(std::move(msg));
   }
 
   // Don't notify if a sync dispatch is in progress, it will wake after finishing.
-  // This might only happen if we started receving messages while `SUBSCRIBE`
-  // is still updating thread local data (see channel_store). We need to make sure its
-  // ack is sent before all other messages.
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:1171]: With ACL messages and checkpoints this is no longer true

File: src/server/cluster/cluster_family.cc [hunk 3]
```diff
@@ -504,11 +504,12 @@ void ClusterFamily::DflyClusterConfig(CmdArgList args, ConnectionContext* cntx)
 
   auto cb = [&](util::ProactorBase* pb) { tl_cluster_config = new_config; };
   server_family_->service().proactor_pool().AwaitFiberOnAll(std::move(cb));
-
   DCHECK(tl_cluster_config != nullptr);
 
-  SlotSet after = tl_cluster_config->GetOwnedSlots();
+  bool success = server_family_->AwaitCurrentDispatches(absl::Seconds(1), cntx->conn());
+  // TODO... and what to do?
```
Comment (dranikpg) [inline src/server/cluster/cluster_family.cc:510]: @chakaz, @adiholden 

Should there be a timeout during a cluster config push? (i.e. waiting for a command that is potentially still using the old config)

If so, this is really only likely with blocking commands (or someone runs scripts that take 1s), should we cancel blocking commands?

File: src/facade/dragonfly_listener.cc [hunk 4]
```diff
@@ -234,26 +234,31 @@ void Listener::PreAcceptLoop(util::ProactorBase* pb) {
...
-    auto cb = [&any_connection_dispatching, &filter](unsigned thread_index,
-                                                     util::Connection* conn) {
-      if (filter(conn) && static_cast<Connection*>(conn)->IsCurrentlyDispatching()) {
-        any_connection_dispatching.store(true);
+  // Fill blocking counter with ongoing dispatches
+  fb2::BlockingCounter bc{0};
+  this->TraverseConnections([bc, issuer](unsigned thread_index, util::Connection* conn) {
+    if (conn != issuer)
+      static_cast<Connection*>(conn)->AwaitCurrentDispatches(bc);
+  });
```
Comment (dranikpg) [inline src/facade/dragonfly_listener.cc:245]: An issue here is that by the time this callback gets to run on some thread, it might have already advanced to some other command, so we might be falsly waiting for one that actually doesn't have any old version of the state

But to solve this issue we'd need to track the dispatch time of each command as well as the time the global state gets updated (including on each thread if it pushes TL values)

**Comment by dranikpg** [comment]
I've narrowed down the CI failure locally to this, still looking into it 🤔🤔🤔 
image

File: src/server/cluster/cluster_family.cc (hunk 3 above)
Comment (chakaz) [reply src/server/cluster/cluster_family.cc:510]: That's a very good question.
Blocking commands specifically can never end, so I guess we should be cancelling them. Not sure about 1s though, it might be a bit too short?
@adiholden wdyt?

**Comment by chakaz** [comment]
> I've narrowed down the CI failure locally to this, still looking into it 🤔🤔🤔 image

Where is this check? Is it new? I can't find it. Where do we assume that we own at least 1 slot? This assumption should not be generally applied, it's OK I guess to not own slots, at least temporarily

**Comment by dranikpg** [comment]
> > I've narrowed down the CI failure locally to this, still looking into it 🤔🤔🤔 image
> 
> Where is this check? Is it new? I can't find it. Where do we assume that we own at least 1 slot? This assumption should not be generally applied, it's OK I guess to not own slots, at least temporarily

Yes, it's a local run version, I'm just trying to find the cause of the very specific failure that occurred on CI. It fails once a couple of hundred runs. The CI output shows that the db is empty and I discovered that for some reason the pushed config is empty with this dcheck (one of my assumptions)

File: src/facade/dragonfly_connection.h [hunk 5]
```diff
@@ -114,13 +121,18 @@ class Connection : public util::Connection {
   using PipelineMessagePtr = std::unique_ptr<PipelineMessage, MessageDeleter>;
   using PubMessagePtr = std::unique_ptr<PubMessage, MessageDeleter>;
 
+  // Variant wrapper around different message types
   struct MessageHandle {
     size_t UsedMemory() const;  // How much bytes this handle takes up in total.
 
+    // Intrusive messages put themselves at the front of the queue, but only after all other
+    // intrusive ones. Used for quick transfer or control / update messages.
+    bool IsIntrusive() const;
```
Comment (romange) [inline src/facade/dragonfly_connection.h:130]: I think `IsControlMsg()` is a better name, because "intrusiveness" is an implementation detail.

File: src/facade/dragonfly_connection.cc [hunk 6]
```diff
@@ -1138,37 +1163,26 @@ void Connection::SendAsync(MessageHandle msg) {
   DCHECK(owner());
   DCHECK_EQ(ProactorBase::me(), socket_->proactor());
 
-  if (cc_->conn_closing)
+  if (cc_->conn_closing && !msg.IsIntrusive())
```
Comment (romange) [inline src/facade/dragonfly_connection.cc:1166]: worth adding a comment why this check

File: src/facade/dragonfly_connection.cc [hunk 7]
```diff
@@ -190,35 +191,35 @@ size_t Connection::PipelineMessage::StorageCapacity() const {
...
-
-  template <typename T, typename D> size_t operator()(const unique_ptr<T, D>& ptr) {
-    return operator()(*ptr.get());
-  }
-};
-
-template <class... Ts> Overloaded(Ts...) -> Overloaded<Ts...>;
-
 size_t Connection::MessageHandle::UsedMemory() const {
-  // TODO: don't count inline size
```
Comment (romange) [inline src/facade/dragonfly_connection.cc:204]: I know it's a draft, but just in case: better separate all the refactoring and the functionality change into separate commits

File: src/server/cluster/cluster_family.cc (hunk 3 above)
Comment (romange) [reply src/server/cluster/cluster_family.cc:510]: we should cancel blocking commands with an INTERRUPTED error or similar.

File: src/server/dflycmd.cc [hunk 8]
```diff
@@ -413,14 +413,11 @@ void DflyCmd::TakeOver(CmdArgList args, ConnectionContext* cntx) {
   absl::Time start = absl::Now();
   AggregateStatus status;
 
+  sf_->CancelBlockingCommands();
+
   // We need to await for all dispatches to finish: Otherwise a transaction might be scheduled
   // after this function exits but before the actual shutdown.
-  sf_->CancelBlockingCommands();
```
Comment (romange) [inline src/server/dflycmd.cc:418]: following up on the previous question - it's been done already, no?

Comment (dranikpg) [reply src/server/dflycmd.cc:418]: No, it's for takeover

File: src/facade/dragonfly_connection.h (hunk 5 above)
Comment (dranikpg) [reply src/facade/dragonfly_connection.h:130]: I agree, however `MigrationRequestMessage` is a control message, but it's not intrusive 🤔 The only "implementation" detail we expose here is that we have a linear order of messages

File: src/server/cluster/cluster_family.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/cluster/cluster_family.cc:510]: Another issue is that we also might be not able to revert the changes, if another endless command starts running on a new slot.

For example, If I start `EVAL while true; do ... ; end` and we fail to switch the config, we will also fail to revert it back (even with a sophisticated solution that tracks slots, because I can start the same script on different slots now). It's an in practice impossible scenario, but still

Comment (dranikpg) [reply src/server/cluster/cluster_family.cc:510]: We could just be doing a client pause actually at the start, thought with most cases with simple commands its not necessary to interrupt traffic at all to do a config switch

File: src/facade/dragonfly_connection.h (hunk 5 above)
Comment (romange) [reply src/facade/dragonfly_connection.h:130]: why the order is important for your logic?

File: src/facade/dragonfly_connection.h [hunk 9]
```diff
@@ -105,6 +94,24 @@ class Connection : public util::Connection {
...
+  struct AclUpdateMessage {
+    std::vector<std::string> username;
+    std::vector<uint32_t> categories;
+    std::vector<std::vector<uint64_t>> commands;
+  };
+
+  // Migration request message, the dispatch fiber stops to give way for thread migration.
+  struct MigrationRequestMessage {};
+
+  // Checkpoint message, decrements counter when processed.
```
Comment (romange) [inline src/facade/dragonfly_connection.h:110]: please expand when (how?) checkpoint message is used.

Comment (romange) [reply src/facade/dragonfly_connection.h:110]: i.e. "dummy message used to make sure that all the currently queues messages are processed in the queue"

File: src/facade/dragonfly_connection.h (hunk 5 above)
Comment (dranikpg) [reply src/facade/dragonfly_connection.h:130]: because ACL updates and checkpoint need to be applied immediately (after the next command finished), where as the migration request wants to be last explicitly to make sure all commands before it have been processed

File: src/facade/dragonfly_connection.cc [hunk 10]
```diff
@@ -1116,6 +1132,14 @@ void Connection::SendAclUpdateAsync(AclUpdateMessage msg) {
   SendAsync({std::move(msg)});
 }
 
+void Connection::AwaitCurrentDispatches(fb2::BlockingCounter bc) {
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/facade/dragonfly_connection.cc]: True, maybe SendCheckpoint akin to others is a better name, or maybe RegisterCheckpoint

File: src/facade/dragonfly_connection.cc [hunk 11]
```diff
@@ -1016,10 +1029,13 @@ void Connection::DispatchFiber(util::FiberSocketBase* peer) {
...
   }
 
-  cc_->conn_closing = true;
+  if (builder->GetError())
+    cc_->conn_closing = true;
 
   // Recycle messages even from disconnecting client to keep properly track of memory stats
   for (auto& msg : dispatch_q_) {
+    if (msg.IsIntrusive())
+      visit(dispatch_op, msg.handle);  // to not miss checkpoints
```
Comment (adiholden) [inline src/facade/dragonfly_connection.cc:1038]: maybe it is safer to add a fiber guard above here to make sure intrusive handlers do not preemt

File: src/facade/dragonfly_connection.cc [hunk 12]
```diff
@@ -1016,10 +1029,13 @@ void Connection::DispatchFiber(util::FiberSocketBase* peer) {
     queue_backpressure_->ec.notify();
   }
 
-  cc_->conn_closing = true;
+  if (builder->GetError())
```
Comment (adiholden) [inline src/facade/dragonfly_connection.cc:1032]: why this change?

File: src/facade/dragonfly_connection.cc (hunk 11 above)
Comment (dranikpg) [reply src/facade/dragonfly_connection.cc:1038]: Will do, but in theory there shouldn't be a problem

File: src/facade/dragonfly_connection.cc (hunk 12 above)
Comment (dranikpg) [reply src/facade/dragonfly_connection.cc:1032]: Will remove, I eventually kept the return below ShouldEndDispatchFiber. Otherwise breaking from the loop does not always equal closing the connection,(migration request)

File: src/facade/dragonfly_connection.cc (hunk 11 above)
Comment (adiholden) [reply src/facade/dragonfly_connection.cc:1038]: We are iterating the dispatch_q_ here, if we preempt we could insert dispatch_q_ from another fiber

**Review by adiholden — APPROVED** [review]
